src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text.Projection;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Venus
+{
+    internal sealed partial class ContainedDocument
+    {
+        public class DocumentServiceProvider : IDocumentServiceProvider
+        {
+            public static readonly IDocumentServiceProvider Instace = new DocumentServiceProvider();
+
+            private readonly SpanMapper _service;
+
+            private DocumentServiceProvider()
+            {
+                _service = new SpanMapper();
+            }
+
+            public TService GetService<TService>() where TService : IDocumentService
+            {
+                if (_service is TService service)
+                {
+                    return service;
+                }
+
+                return default;
+            }
+
+            private class SpanMapper : ISpanMapperService
+            {
+                public async Task<ImmutableArray<DocumentSpan>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken)
+                {
+                    var containedDocument = ContainedDocument.TryGetContainedDocument(document.Id);
+                    if (containedDocument == null)
```

Comment:
How could this ever happen? Seems like we'd want to throw, since something would be most amiss?

Reply:
sure. can make it to throw